New features

ZhNV2 opened the pull request:
1. Whiteboards
2. Smart pop-up notification about found battle
3. Accept/decline battle

Qwsafex committed:
First review of description.

Qwsafex committed:
Added plan and fixed a typo in description.

Qwsafex committed:
Started writing detailed plan

Qwsafex committed:
Added UML diagram.

Qwsafex committed:
Added architecture images

Qwsafex committed:
Added rest sample

Qwsafex committed:
Android project created

ZhNV2 committed:
Chat Activity and connected classes were added

Qwsafex committed:
Added most classes stubs

ZhNV2 committed:
Chat Activity and connected classes were added

ZhNV2 committed:
Now search button moves you to chat activity

Qwsafex committed:
File structure reorganized.

Qwsafex committed:
Added GET request for RequestMaker; added profile info display in ProfileActivity
Added basic serverside ProfileManager

Qwsafex committed:
removed some trash from repo

Qwsafex committed:
Added LoginActivity.
Added simple sign in mechanism on server side.

Qwsafex committed:
Added all HTTP methods to RequestMaker.sendRequest()

Qwsafex committed:
Small fix

ZhNV2 committed:
first version of chat-backend. create_chat, close_chat, get_messages, send_message are now available

Qwsafex committed:
Added SignupActivity.
Minor style fixes on LoginActivity.

Qwsafex committed:
Minor style/spelling fixes

Qwsafex committed:
Message.map_repr: author_id -> authorId

Qwsafex committed:
Added Chatter

ZhNV2 committed:
mentioned problems from pull request were fixed

Qwsafex committed:
Minor style fix

Qwsafex committed:
Merge pull request #1 from BatherGaming/server/chat-backend

first version of chat-backend

Qwsafex committed:
Added signup

signup'
Added signup

# Please enter a commit message to explain why this merge is necessary,
# especially if it merges an updated upstream into a topic branch.
#
# Lines starting with '#' will be ignored, and an empty message aborts
# the commit.

Qwsafex committed:
Merge branch 'master' of https://github.com/BatherGaming/chat-battle
fixed minor conflicts

Qwsafex committed:
Merge pull request #2 from BatherGaming/signup

Added SignupActivity.

ZhNV2 committed:
domain name is moved to request maker; text field in chat becames clear after sending

Qwsafex committed:
Chatter becomes ChatService and looks finished

Qwsafex committed:
Added getMessages() to ChatService

ZhNV2 committed:
battle-searcher

Qwsafex committed:
Added simple battlemaker

ZhNV2 committed:
simple version of battle searcher and modifications of chat activity

Qwsafex committed:
Merged master

Qwsafex committed:
Merged battle_searcher

Qwsafex committed:
Fixed bugs

Qwsafex committed:
Fixed minor bug

Qwsafex committed:
Debug stuff

Qwsafex committed:
Working (?) version

Qwsafex committed:
Bugs fixed. Works just fine.

ZhNV2 committed:
leader is implemented

ZhNV2 committed:
autoscroll

Qwsafex committed:
First steps to remove abundant SearchActivity

Qwsafex committed:
Removed debug toast.

Qwsafex committed:
Fixed stupid bugs

ZhNV2 committed:
first corrections

ZhNV2 committed:
fixed pull request issues

Qwsafex committed:
PEP8 style fixes

Qwsafex committed:
Typo corrected

Qwsafex committed:
Typo corrected

Qwsafex committed:
Renamed ChatActivity to PlayerActivity. Chanded base class of LeaderActivity to AbstractChat.

ZhNV2 committed:
Merge branch 'basic/leader' of https://github.com/BatherGaming/chat-battle into basic/leader

Qwsafex committed:
Added test whiteboard and python

Qwsafex committed:
Removed trash

Qwsafex committed:
Test version of whiteboard. You can draw and send drawings to server.

ZhNV2 committed:
1. spinner while waiting for battle 2. pop-up window in any activity when battle is founded. 3. correct jump when button play was clicked

ZhNV2 committed:
smart searching activity

ZhNV2 committed:
fucking misc.xml now in gitignore!))

Qwsafex committed:
Whiteboards fully implemented

Qwsafex committed:
Added whiteboard button image change to display presence of attached image

ZhNV2 committed:
1.popup dialog when battle is found. 2. correct processing diffirent cases (decline, both silence, etc). 3. Back button in dialog was forbidden

Qwsafex committed:
Removed trash

Qwsafex committed:
Merging #1 into new-features

 new-features

# Please enter a commit message to explain why this merge is necessary,
# especially if it merges an updated upstream into a topic branch.
#
# Lines starting with '#' will be ignored, and an empty message aborts
# the commit.

Qwsafex committed:
Merge 1,2 complete

Qwsafex committed:
Merge of 1,2,3 completed

Qwsafex committed:
fromJSON in Message and Player added. Eraser to WhiteboardActivity added

ZhNV2 committed:
for ILya

Qwsafex committed:
Fixed SearchActivity using constant string instead of resource. Added Russian strings.

ZhNV2 committed:
almost all cases from issue 26 work fine

ZhNV2 committed:
Merge branch 'progress/4' into new-features

ZhNV2 committed:
issue 7 implemented (rating)

ZhNV2 committed:
translation for rating

Qwsafex committed:
Added UTF-8 encoding to sent data allowing for display of non-English messages (issue #11)

ZhNV2 committed:
implemented issue #20(push notifications)

Qwsafex committed:
Made messages appear immidiately (issue #10)

sproshev commented on server/profile_manager.py:
```diff
@@ -0,0 +1,65 @@
+def get_chat_players(chat_id):
+    chat = session.query(Chat).filter_by(id=chat_id).first()
+    if not chat:
+        return {"error": "Chat doesn't exist"}, 400
```
почему не 404?

sproshev commented on server/profile_manager.py:
```diff
@@ -0,0 +1,65 @@
+def add_player(json):
+    if not json:
+        return {"error": "JSON is null"}, 400
+    for parametr in ['login', 'sex', 'password']:
```
parameter

sproshev commented on server/profile_manager.py:
```diff
@@ -0,0 +1,65 @@
+        if parametr not in json:
+            return {"error": "no " + parametr + " in JSON"}, 400
+    for parametr in ['login', 'password']:
+        if json[parametr] == "":
```
`if not json[parameter]`

sproshev commented on server/profile_manager.py:
```diff
@@ -0,0 +1,65 @@
+            return {"error": "no " + parametr + " in JSON"}, 400
+    for parametr in ['login', 'password']:
+        if json[parametr] == "":
+            return {"error": parametr + " should be not empty"}
```
код ошибки

sproshev commented on server/profile_manager.py:
```diff
@@ -0,0 +1,65 @@
+        return {"error": "Login is taken."}, 422  # Unprocessible entity
+
+    password = json.get("password", "")
+    password_hash = hashlib.md5(password.encode("utf-8")).hexdigest()
```
вычисление хеша дублируется --> метод

sproshev commented on server/db.py:
```diff
@@ -0,0 +1,108 @@
+from sqlalchemy.orm import sessionmaker, relationship
+from sqlalchemy import Column, Integer, String, ForeignKey
+from sqlalchemy.ext.declarative import declarative_base
+from sqlalchemy.sql import text
```
unused

sproshev commented on server/db.py:
```diff
@@ -0,0 +1,108 @@
+    rating = Column(Integer)
+
+    messages = relationship("Message", backref="sender")
+    # chats = relationship("Chat", backref="leader") # TODO: not chats but chat
```
почему такое TODO? ведущим он тоже только в 1 чате может быть?

ZhNV2 replied:
Я сам не могу вспомнить, к чему это, просто убрал

sproshev commented on server/db.py:
```diff
@@ -0,0 +1,108 @@
+        return Player(id=message_dict["id"], login=message_dict["login"],
+                      sex=message_dict["sex"], age=message_dict["age"],
+                      chat_id=message_dict["chatId"], status=message_dict["status"],
+                      rating=message_dict["rating"])
```
метод поломан

sproshev commented on server/db.py:
```diff
@@ -0,0 +1,108 @@
+    is_closed = Column(Boolean)
+    leader_id = Column(Integer)
+    winner_id = Column(Integer)
+    accepted = Column(Integer, default=0)
```
это поле про что?

ZhNV2 replied:
сколько игроков уже приняло приглашение вступить в игру

sproshev commented on server/chat_backend.py:
```diff
@@ -0,0 +1,192 @@
+    session.add(chat)
+    for playerId in itertools.chain(players_ids, [leader_id]):
+        player = session.query(Player).filter_by(id=playerId).first()
+        player.chat_id = chat.id
```
вопрос про единственный чат для каждого игрока все еще в силе

понял, игрок в один момент времени может участвовать только в 1 чате

ZhNV2 replied:
да, для игрока в каждый момент времени один чат

sproshev commented on server/chat_backend.py:
```diff
@@ -0,0 +1,192 @@
+def close_chat(leader_id, winner_id):
+    leader = session.query(Player).filter_by(id=leader_id).first()
+    if not leader:
+        return {"error": "player with provided leader_id doesn't exist"}
```
код ошибки

sproshev commented on server/chat_backend.py:
```diff
@@ -0,0 +1,192 @@
+    session.commit()
+    return chat.id
+
+def close(chat, started, winner_id):
```
флаг is_started есть в chat, зачем его передавать?

sproshev commented on server/chat_backend.py:
```diff
@@ -0,0 +1,192 @@
+        player.chat_id = None
+        player.status = "IDLE"
+        if started and player.id != chat.leader_id:
+            player.rating = player.rating + (WINNER_DELTA if player.id == winner_id else LOSER_DELTA)
```
+=

sproshev commented on server/chat_backend.py:
```diff
@@ -0,0 +1,192 @@
+        return {}, 400
+    if json["text"] == "" and ('whiteboard' not in json 
+                               or json['whiteboard'] == ""):
+        return {}, 400
```
можно объединить эти два if'а, если ничего менять не планируется

sproshev commented on server/chat_backend.py:
```diff
@@ -0,0 +1,192 @@
+            or not str(json["chatId"]).isdigit()\
+            or not str(json["authorId"]).isdigit():
+        return {}, 400
+    if json["text"] == "" and ('whiteboard' not in json 
```
whiteboard что за штука? почему в сообщении должна передаваться?

ZhNV2 replied:
whiteboard - это нарисованная на клиенте картинка, которая передается на сервер

sproshev commented on server/chat_backend.py:
```diff
@@ -0,0 +1,192 @@
+
+def get_leader(chat_id):
+    chat = session.query(Chat).filter_by(id=chat_id).first()
+    if not leader:
```
тоже сломано

sproshev commented on server/chat_backend.py:
```diff
@@ -0,0 +1,192 @@
+    chat = session.query(Chat).filter_by(id=chat_id).first()
+    if not leader:
+        return {"error": "Chat doesn't exist"}, 400
+    return {"leader": chat.leader_id}
```
200

sproshev commented on server/chat_backend.py:
```diff
@@ -0,0 +1,192 @@
+    if chat.is_closed or chat.is_started:
+        return
+    if (datetime.datetime.now() - chat.creation_time).total_seconds() > 25:
+        close(chat, False, -1)
```
из названия не очень понятно, зачем этот метод, что верифицирует и почему закрывает чат

ZhNV2 replied:
это на тот случай, если после создания баттла все игроки резко уйдут из игры и тогда чат так и будет висеть в непонятном открытом состоянии. А verify проверяет, что если чат открыл слишком долго, его уже не надо создавать. В этой реализации чат закроется, как только первый из игроков снова войдет в игру и при первом обращении вызовется verify

Qwsafex committed:
Fixed PlayerActivity crashes. Added VIBRATE permission. Made chat wait for player ids to load

sproshev commented on server/chat_backend.py:
```diff
@@ -0,0 +1,192 @@
+    elif chat.is_started:
+        return {"result": "running"}, 200
+    else:
+        return {"result": "waiting"}, 200
```
не очень хорошо, что в метода смешался как статус чата, так и статус игрока в чате

ZhNV2 replied:
Так сделано, потому что не очень хочется делать 2 запроса к серверу вместо одного

sproshev commented on server/chat_backend.py:
```diff
@@ -0,0 +1,192 @@
+    chat = session.query(Chat).filter_by(id=player.chat_id).first()
+    verify(chat)
+   
+    chat.accepted += 1
```
аа, понял, это количество тех, кто "вступил" в чат. надо подумать над именем

sproshev commented on server/chat_backend.py:
```diff
@@ -0,0 +1,192 @@
+    verify(chat)
+   
+    chat.accepted += 1
+    if chat.accepted == len(chat.players):
```
chat.players это что-то наподобие "приглашенных"?

ZhNV2 replied:
Да, если чат откроется, то они будут его игроками

sproshev commented on server/chat_backend.py:
```diff
@@ -0,0 +1,192 @@
+        player.status = "IDLE"
+        if started and player.id != chat.leader_id:
+            player.rating = player.rating + (WINNER_DELTA if player.id == winner_id else LOSER_DELTA)
+    session.commit()
```
чат могут закрыть несколько раз? рейтинг тогда каждый раз будет пересчитываться (

sproshev commented on server/chat_backend.py:
```diff
@@ -0,0 +1,192 @@
+    with open(location, "rb") as f:
+        whiteboard = f.read()
+        return base64.b64encode(whiteboard)
+    return ""
```
недостижимая инструкция

sproshev commented on server/battlemaker.py:
```diff
@@ -0,0 +1,78 @@
+def read_queues_from_file():
+    queues = []
+    with open(queue_filename, "r") as queue_file:
+        queues = [queue_file.readline()[:-1].split(",") for i in range(2)]
```
`for _ in range(2)`

sproshev commented on server/battlemaker.py:
```diff
@@ -0,0 +1,78 @@
+
+def write_queues_in_file(players_queue, leaders_queue):
+    with open(queue_filename, "w") as queue_file:
+        queue_file.write(",".join(map(str, players_queue)) + '\n')
```
лучше строки не конкатенировать, а "\n" отдельно записать

sproshev commented on server/battlemaker.py:
```diff
@@ -0,0 +1,78 @@
+        if player_id in leaders_queue:
+            leaders_queue.remove(player_id)
+        players_queue.append(player_id)
+        player.status = "IN_QUEUE_AS_PLAYER"
```
в питоне есть енамы, только не знаю, как их в бд класть

sproshev commented on server/battlemaker.py:
```diff
@@ -0,0 +1,78 @@
+
+    players_queue, leaders_queue = read_queues_from_file()
+
+    player = session.query(Player).filter_by(id=player_id).first()
```
один запрос производится два раза

sproshev reviewed (commented):
сервер посмотрел

ZhNV2 commented:
до свидания17.12.2016, 22:45, "Semyon Proshev" <notifications@github.com>:@sproshev commented on this pull request.

сервер посмотрел

In server/profile_manager.py:
> +
+from db import session, Player, Chat
+from sqlalchemy import and_
+
+BASIC_RATING = 1000
+
+
+def get_players():
+    players = session.query(Player).all()
+    return list(map(Player.toDict, players)), 200
+
+
+def get_chat_players(chat_id):
+    chat = session.query(Chat).filter_by(id=chat_id).first()
+    if not chat:
+        return {"error": "Chat doesn't exist"}, 400

почему не 404?



In server/profile_manager.py:
> +                                  Player.id != chat.leader_id))\
+                     .all()
+    return list(map(Player.toDict, players)), 200
+
+
+def get_player(player_id):
+    player = session.query(Player).filter_by(id=player_id).first()
+    if not player:
+        return {}, 404
+    return player.toDict(), 200
+
+
+def add_player(json):
+    if not json:
+        return {"error": "JSON is null"}, 400
+    for parametr in ['login', 'sex', 'password']:

parameter



In server/profile_manager.py:
> +
+def get_player(player_id):
+    player = session.query(Player).filter_by(id=player_id).first()
+    if not player:
+        return {}, 404
+    return player.toDict(), 200
+
+
+def add_player(json):
+    if not json:
+        return {"error": "JSON is null"}, 400
+    for parametr in ['login', 'sex', 'password']:
+        if parametr not in json:
+            return {"error": "no " + parametr + " in JSON"}, 400
+    for parametr in ['login', 'password']:
+        if json[parametr] == "":

if not json[parameter]



In server/profile_manager.py:
> +def get_player(player_id):
+    player = session.query(Player).filter_by(id=player_id).first()
+    if not player:
+        return {}, 404
+    return player.toDict(), 200
+
+
+def add_player(json):
+    if not json:
+        return {"error": "JSON is null"}, 400
+    for parametr in ['login', 'sex', 'password']:
+        if parametr not in json:
+            return {"error": "no " + parametr + " in JSON"}, 400
+    for parametr in ['login', 'password']:
+        if json[parametr] == "":
+            return {"error": parametr + " should be not empty"}

код ошибки



In server/profile_manager.py:
> +
+def add_player(json):
+    if not json:
+        return {"error": "JSON is null"}, 400
+    for parametr in ['login', 'sex', 'password']:
+        if parametr not in json:
+            return {"error": "no " + parametr + " in JSON"}, 400
+    for parametr in ['login', 'password']:
+        if json[parametr] == "":
+            return {"error": parametr + " should be not empty"}
+
+    if session.query(Player).filter_by(login=json['login']).first():
+        return {"error": "Login is taken."}, 422  # Unprocessible entity
+
+    password = json.get("password", "")
+    password_hash = hashlib.md5(password.encode("utf-8")).hexdigest()

вычисление хеша дублируется --> метод



In server/db.py:
> @@ -0,0 +1,108 @@
+import datetime
+from sqlalchemy import create_engine
+
+from sqlalchemy.orm import sessionmaker, relationship
+from sqlalchemy import Column, Integer, String, ForeignKey
+from sqlalchemy.ext.declarative import declarative_base
+from sqlalchemy.sql import text

unused



In server/db.py:
> +
+
+class Player(Base):
+    __tablename__ = 'players'
+
+    id = Column(Integer, primary_key=True)
+    login = Column(String, nullable=False)
+    age = Column(Integer)
+    password_hash = Column(String)
+    sex = Column(String)  # 'MALE'/'FEMALE'
+    chat_id = Column(Integer, ForeignKey('chats.id'))
+    status = Column(String)   
+    rating = Column(Integer)
+
+    messages = relationship("Message", backref="sender")
+    # chats = relationship("Chat", backref="leader") # TODO: not chats but chat

почему такое TODO? ведущим он тоже только в 1 чате может быть?



In server/db.py:
> +
+    def toDict(self):
+        return {"id": self.id,
+                "login": self.login,
+                "sex": self.sex,
+                "age": self.age,
+                "chatId": self.chat_id,
+                "status": self.status,
+                "rating": self.rating}
+
+    @staticmethod
+    def fromDict(player_dict):
+        return Player(id=message_dict["id"], login=message_dict["login"],
+                      sex=message_dict["sex"], age=message_dict["age"],
+                      chat_id=message_dict["chatId"], status=message_dict["status"],
+                      rating=message_dict["rating"])

метод поломан



In server/db.py:
> +                      chat_id=message_dict["chatId"], status=message_dict["status"],
+                      rating=message_dict["rating"])
+
+
+class Chat(Base):
+    __tablename__ = 'chats'
+
+    id = Column(Integer, primary_key=True)
+    creation_time = Column(DateTime)
+    type = Column(Integer)
+
+    is_started = Column(Boolean, default=False)
+    is_closed = Column(Boolean)
+    leader_id = Column(Integer)
+    winner_id = Column(Integer)
+    accepted = Column(Integer, default=0)

это поле про что?



In server/chat_backend.py:
> +
+from db import session, Message, Player, Chat, datetime
+
+WHITEBOARD_FOLDER = 'whiteboards'
+DOMAIN_NAME = "qwsafex.pythonanywhere.com"  
+
+WINNER_DELTA = 1
+LOSER_DELTA = -1
+
+def create_chat(type, players_ids, leader_id):
+    chat = Chat(type=type, creation_time=datetime.datetime.now(),
+                is_closed=False, leader_id=leader_id, accepted=0)
+    session.add(chat)
+    for playerId in itertools.chain(players_ids, [leader_id]):
+        player = session.query(Player).filter_by(id=playerId).first()
+        player.chat_id = chat.id

вопрос про единственный чат для каждого игрока все еще в силе
понял, игрок в один момент времени может участвовать только в 1 чате



In server/chat_backend.py:
> +    session.commit()
+    return chat.id
+
+def close(chat, started, winner_id):
+    chat.is_closed = True
+    for player in chat.players:
+        player.chat_id = None
+        player.status = "IDLE"
+        if started and player.id != chat.leader_id:
+            player.rating = player.rating + (WINNER_DELTA if player.id == winner_id else LOSER_DELTA)
+    session.commit()
+
+def close_chat(leader_id, winner_id):
+    leader = session.query(Player).filter_by(id=leader_id).first()
+    if not leader:
+        return {"error": "player with provided leader_id doesn't exist"}

код ошибки



In server/chat_backend.py:
> +
+WINNER_DELTA = 1
+LOSER_DELTA = -1
+
+def create_chat(type, players_ids, leader_id):
+    chat = Chat(type=type, creation_time=datetime.datetime.now(),
+                is_closed=False, leader_id=leader_id, accepted=0)
+    session.add(chat)
+    for playerId in itertools.chain(players_ids, [leader_id]):
+        player = session.query(Player).filter_by(id=playerId).first()
+        player.chat_id = chat.id
+        player.status = "CHATTING_AS_LEADER" if player.id == leader_id else "CHATTING_AS_PLAYER"
+    session.commit()
+    return chat.id
+
+def close(chat, started, winner_id):

флаг is_started есть в chat, зачем его передавать?



In server/chat_backend.py:
> +                is_closed=False, leader_id=leader_id, accepted=0)
+    session.add(chat)
+    for playerId in itertools.chain(players_ids, [leader_id]):
+        player = session.query(Player).filter_by(id=playerId).first()
+        player.chat_id = chat.id
+        player.status = "CHATTING_AS_LEADER" if player.id == leader_id else "CHATTING_AS_PLAYER"
+    session.commit()
+    return chat.id
+
+def close(chat, started, winner_id):
+    chat.is_closed = True
+    for player in chat.players:
+        player.chat_id = None
+        player.status = "IDLE"
+        if started and player.id != chat.leader_id:
+            player.rating = player.rating + (WINNER_DELTA if player.id == winner_id else LOSER_DELTA)

+=



In server/chat_backend.py:
> +    filename = whiteboard_location(base_filename)
+    with open(filename, "wb") as file:
+        file.write(whiteboard_body)
+    return base_filename
+
+
+def send_message(json):
+    if not json or 'authorId' not in json\
+            or 'text' not in json\
+            or 'chatId' not in json\
+            or not str(json["chatId"]).isdigit()\
+            or not str(json["authorId"]).isdigit():
+        return {}, 400
+    if json["text"] == "" and ('whiteboard' not in json 
+                               or json['whiteboard'] == ""):
+        return {}, 400

можно объединить эти два if'а, если ничего менять не планируется



In server/chat_backend.py:
> +    whiteboard_body = base64.b64decode(whiteboard_body)
+    base_filename = unique_filename()
+    filename = whiteboard_location(base_filename)
+    with open(filename, "wb") as file:
+        file.write(whiteboard_body)
+    return base_filename
+
+
+def send_message(json):
+    if not json or 'authorId' not in json\
+            or 'text' not in json\
+            or 'chatId' not in json\
+            or not str(json["chatId"]).isdigit()\
+            or not str(json["authorId"]).isdigit():
+        return {}, 400
+    if json["text"] == "" and ('whiteboard' not in json 

whiteboard что за штука? почему в сообщении должна передаваться?



In server/chat_backend.py:
> +    session.commit()
+    return message.toDict(), 200
+
+
+def get_messages(chat_id, num):
+    messages = session.query(Message).\
+                        filter_by(chat_id=chat_id).\
+                        order_by(Message.time).\
+                        offset(num).\
+                        all()
+    return list(map(Message.toDict, messages)), 200
+
+
+def get_leader(chat_id):
+    chat = session.query(Chat).filter_by(id=chat_id).first()
+    if not leader:

тоже сломано



In server/chat_backend.py:
> +
+
+def get_messages(chat_id, num):
+    messages = session.query(Message).\
+                        filter_by(chat_id=chat_id).\
+                        order_by(Message.time).\
+                        offset(num).\
+                        all()
+    return list(map(Message.toDict, messages)), 200
+
+
+def get_leader(chat_id):
+    chat = session.query(Chat).filter_by(id=chat_id).first()
+    if not leader:
+        return {"error": "Chat doesn't exist"}, 400
+    return {"leader": chat.leader_id}

200



In server/chat_backend.py:
> +                        all()
+    return list(map(Message.toDict, messages)), 200
+
+
+def get_leader(chat_id):
+    chat = session.query(Chat).filter_by(id=chat_id).first()
+    if not leader:
+        return {"error": "Chat doesn't exist"}, 400
+    return {"leader": chat.leader_id}
+
+
+def verify(chat):
+    if chat.is_closed or chat.is_started:
+        return
+    if (datetime.datetime.now() - chat.creation_time).total_seconds() > 25:
+        close(chat, False, -1)

из названия не очень понятно, зачем этот метод, что верифицирует и почему закрывает чат



In server/chat_backend.py:
> +        return {"error": "Player doesn't exist"}, 400
+    if not chat.is_closed and player not in chat.players:
+        return {"error": "You are not in required chat"}, 400
+    if chat.is_closed:
+        if not chat.is_started:
+            return {"result": "won't start"}, 200
+        elif chat.leader_id == player.id:
+            return {"result": "leader"}, 200
+        elif chat.winner_id == player.id:
+            return {"result": "winner", "rating": player.rating}, 200
+        else:
+            return {"result": "loser", "rating": player.rating}, 200
+    elif chat.is_started:
+        return {"result": "running"}, 200
+    else:
+        return {"result": "waiting"}, 200

не очень хорошо, что в метода смешался как статус чата, так и статус игрока в чате



In server/chat_backend.py:
> +        return {"result": "running"}, 200
+    else:
+        return {"result": "waiting"}, 200
+
+
+def accept(player_id):
+    print('accepted', player_id)
+    player = session.query(Player).filter_by(id=player_id).first()
+    if player is None:
+        return {"error": "Player doesn't exist"}, 400
+    if player.chat_id is None:
+        return {"error": "Player is not in chat"}, 400
+    chat = session.query(Chat).filter_by(id=player.chat_id).first()
+    verify(chat)
+   
+    chat.accepted += 1

аа, понял, это количество тех, кто "вступил" в чат. надо подумать над именем



In server/chat_backend.py:
> +    else:
+        return {"result": "waiting"}, 200
+
+
+def accept(player_id):
+    print('accepted', player_id)
+    player = session.query(Player).filter_by(id=player_id).first()
+    if player is None:
+        return {"error": "Player doesn't exist"}, 400
+    if player.chat_id is None:
+        return {"error": "Player is not in chat"}, 400
+    chat = session.query(Chat).filter_by(id=player.chat_id).first()
+    verify(chat)
+   
+    chat.accepted += 1
+    if chat.accepted == len(chat.players):

chat.players это что-то наподобие "приглашенных"?



In server/chat_backend.py:
> +    session.add(chat)
+    for playerId in itertools.chain(players_ids, [leader_id]):
+        player = session.query(Player).filter_by(id=playerId).first()
+        player.chat_id = chat.id
+        player.status = "CHATTING_AS_LEADER" if player.id == leader_id else "CHATTING_AS_PLAYER"
+    session.commit()
+    return chat.id
+
+def close(chat, started, winner_id):
+    chat.is_closed = True
+    for player in chat.players:
+        player.chat_id = None
+        player.status = "IDLE"
+        if started and player.id != chat.leader_id:
+            player.rating = player.rating + (WINNER_DELTA if player.id == winner_id else LOSER_DELTA)
+    session.commit()

чат могут закрыть несколько раз? рейтинг тогда каждый раз будет пересчитываться (



In server/chat_backend.py:
> +    return {}, 200
+
+def get_chat(chat_id):
+    chat = session.query(Chat).filter_by(id=chat_id).first()
+    return chat.toDict(), 200
+
+
+
+    
+
+def get_whiteboard(whiteboard_tag):
+    location = whiteboard_location(whiteboard_tag)
+    with open(location, "rb") as f:
+        whiteboard = f.read()
+        return base64.b64encode(whiteboard)
+    return ""

недостижимая инструкция



In server/battlemaker.py:
> @@ -0,0 +1,78 @@
+from db import Player, session
+from chat_backend import create_chat
+
+queue_filename = "queue.txt"
+player_amount = 1
+
+
+# Make sure file is present
+open(queue_filename, "a").close()
+
+
+def read_queues_from_file():
+    queues = []
+    with open(queue_filename, "r") as queue_file:
+        queues = [queue_file.readline()[:-1].split(",") for i in range(2)]

for _ in range(2)



In server/battlemaker.py:
> +# Make sure file is present
+open(queue_filename, "a").close()
+
+
+def read_queues_from_file():
+    queues = []
+    with open(queue_filename, "r") as queue_file:
+        queues = [queue_file.readline()[:-1].split(",") for i in range(2)]
+    queues = [list(map(int, filter(lambda s: str(s).isdigit(), queue)))
+              for queue in queues]
+    return queues[0], queues[1]
+
+
+def write_queues_in_file(players_queue, leaders_queue):
+    with open(queue_filename, "w") as queue_file:
+        queue_file.write(",".join(map(str, players_queue)) + '\n')

лучше строки не конкатенировать, а "\n" отдельно записать



In server/battlemaker.py:
> +    player = session.query(Player).filter_by(id=player_id).first()
+    if not player:
+        return {"error": "There's no such player."}, 400
+
+    if player.chat_id:
+        return {"error": "Player's already chatting."}, 400
+
+    players_queue, leaders_queue = read_queues_from_file()
+
+    if role == "player":
+        if player_id in players_queue:
+            return {"error": "Player's already in queue."}, 400
+        if player_id in leaders_queue:
+            leaders_queue.remove(player_id)
+        players_queue.append(player_id)
+        player.status = "IN_QUEUE_AS_PLAYER"

в питоне есть енамы, только не знаю, как их в бд класть



In server/battlemaker.py:
> +        for i in range(player_amount):
+            players.append(players_queue.pop(0))
+        create_chat(0, players, leaders_queue.pop(0))
+
+    write_queues_in_file(players_queue, leaders_queue)
+
+    return {}, 200
+
+
+def del_player_from_queue(player_id):
+    if not session.query(Player).filter_by(id=player_id).first():
+        return {"error": "There's no such player."}, 400
+
+    players_queue, leaders_queue = read_queues_from_file()
+
+    player = session.query(Player).filter_by(id=player_id).first()

один запрос производится два раза

—You are receiving this because you authored the thread.Reply to this email directly, view it on GitHub, or mute the thread.

Qwsafex committed:
Warnings fixed

ZhNV2 commented:
Если вы получили сообщение с "до свидания", то я прошу прощения, это случайно17.12.2016, 22:45, "Semyon Proshev" <notifications@github.com>:@sproshev commented on this pull request.

сервер посмотрел

In server/profile_manager.py:
> +
+from db import session, Player, Chat
+from sqlalchemy import and_
+
+BASIC_RATING = 1000
+
+
+def get_players():
+    players = session.query(Player).all()
+    return list(map(Player.toDict, players)), 200
+
+
+def get_chat_players(chat_id):
+    chat = session.query(Chat).filter_by(id=chat_id).first()
+    if not chat:
+        return {"error": "Chat doesn't exist"}, 400

почему не 404?



In server/profile_manager.py:
> +                                  Player.id != chat.leader_id))\
+                     .all()
+    return list(map(Player.toDict, players)), 200
+
+
+def get_player(player_id):
+    player = session.query(Player).filter_by(id=player_id).first()
+    if not player:
+        return {}, 404
+    return player.toDict(), 200
+
+
+def add_player(json):
+    if not json:
+        return {"error": "JSON is null"}, 400
+    for parametr in ['login', 'sex', 'password']:

parameter



In server/profile_manager.py:
> +
+def get_player(player_id):
+    player = session.query(Player).filter_by(id=player_id).first()
+    if not player:
+        return {}, 404
+    return player.toDict(), 200
+
+
+def add_player(json):
+    if not json:
+        return {"error": "JSON is null"}, 400
+    for parametr in ['login', 'sex', 'password']:
+        if parametr not in json:
+            return {"error": "no " + parametr + " in JSON"}, 400
+    for parametr in ['login', 'password']:
+        if json[parametr] == "":

if not json[parameter]



In server/profile_manager.py:
> +def get_player(player_id):
+    player = session.query(Player).filter_by(id=player_id).first()
+    if not player:
+        return {}, 404
+    return player.toDict(), 200
+
+
+def add_player(json):
+    if not json:
+        return {"error": "JSON is null"}, 400
+    for parametr in ['login', 'sex', 'password']:
+        if parametr not in json:
+            return {"error": "no " + parametr + " in JSON"}, 400
+    for parametr in ['login', 'password']:
+        if json[parametr] == "":
+            return {"error": parametr + " should be not empty"}

код ошибки



In server/profile_manager.py:
> +
+def add_player(json):
+    if not json:
+        return {"error": "JSON is null"}, 400
+    for parametr in ['login', 'sex', 'password']:
+        if parametr not in json:
+            return {"error": "no " + parametr + " in JSON"}, 400
+    for parametr in ['login', 'password']:
+        if json[parametr] == "":
+            return {"error": parametr + " should be not empty"}
+
+    if session.query(Player).filter_by(login=json['login']).first():
+        return {"error": "Login is taken."}, 422  # Unprocessible entity
+
+    password = json.get("password", "")
+    password_hash = hashlib.md5(password.encode("utf-8")).hexdigest()

вычисление хеша дублируется --> метод



In server/db.py:
> @@ -0,0 +1,108 @@
+import datetime
+from sqlalchemy import create_engine
+
+from sqlalchemy.orm import sessionmaker, relationship
+from sqlalchemy import Column, Integer, String, ForeignKey
+from sqlalchemy.ext.declarative import declarative_base
+from sqlalchemy.sql import text

unused



In server/db.py:
> +
+
+class Player(Base):
+    __tablename__ = 'players'
+
+    id = Column(Integer, primary_key=True)
+    login = Column(String, nullable=False)
+    age = Column(Integer)
+    password_hash = Column(String)
+    sex = Column(String)  # 'MALE'/'FEMALE'
+    chat_id = Column(Integer, ForeignKey('chats.id'))
+    status = Column(String)   
+    rating = Column(Integer)
+
+    messages = relationship("Message", backref="sender")
+    # chats = relationship("Chat", backref="leader") # TODO: not chats but chat

почему такое TODO? ведущим он тоже только в 1 чате может быть?



In server/db.py:
> +
+    def toDict(self):
+        return {"id": self.id,
+                "login": self.login,
+                "sex": self.sex,
+                "age": self.age,
+                "chatId": self.chat_id,
+                "status": self.status,
+                "rating": self.rating}
+
+    @staticmethod
+    def fromDict(player_dict):
+        return Player(id=message_dict["id"], login=message_dict["login"],
+                      sex=message_dict["sex"], age=message_dict["age"],
+                      chat_id=message_dict["chatId"], status=message_dict["status"],
+                      rating=message_dict["rating"])

метод поломан



In server/db.py:
> +                      chat_id=message_dict["chatId"], status=message_dict["status"],
+                      rating=message_dict["rating"])
+
+
+class Chat(Base):
+    __tablename__ = 'chats'
+
+    id = Column(Integer, primary_key=True)
+    creation_time = Column(DateTime)
+    type = Column(Integer)
+
+    is_started = Column(Boolean, default=False)
+    is_closed = Column(Boolean)
+    leader_id = Column(Integer)
+    winner_id = Column(Integer)
+    accepted = Column(Integer, default=0)

это поле про что?



In server/chat_backend.py:
> +
+from db import session, Message, Player, Chat, datetime
+
+WHITEBOARD_FOLDER = 'whiteboards'
+DOMAIN_NAME = "qwsafex.pythonanywhere.com"  
+
+WINNER_DELTA = 1
+LOSER_DELTA = -1
+
+def create_chat(type, players_ids, leader_id):
+    chat = Chat(type=type, creation_time=datetime.datetime.now(),
+                is_closed=False, leader_id=leader_id, accepted=0)
+    session.add(chat)
+    for playerId in itertools.chain(players_ids, [leader_id]):
+        player = session.query(Player).filter_by(id=playerId).first()
+        player.chat_id = chat.id

вопрос про единственный чат для каждого игрока все еще в силе
понял, игрок в один момент времени может участвовать только в 1 чате



In server/chat_backend.py:
> +    session.commit()
+    return chat.id
+
+def close(chat, started, winner_id):
+    chat.is_closed = True
+    for player in chat.players:
+        player.chat_id = None
+        player.status = "IDLE"
+        if started and player.id != chat.leader_id:
+            player.rating = player.rating + (WINNER_DELTA if player.id == winner_id else LOSER_DELTA)
+    session.commit()
+
+def close_chat(leader_id, winner_id):
+    leader = session.query(Player).filter_by(id=leader_id).first()
+    if not leader:
+        return {"error": "player with provided leader_id doesn't exist"}

код ошибки



In server/chat_backend.py:
> +
+WINNER_DELTA = 1
+LOSER_DELTA = -1
+
+def create_chat(type, players_ids, leader_id):
+    chat = Chat(type=type, creation_time=datetime.datetime.now(),
+                is_closed=False, leader_id=leader_id, accepted=0)
+    session.add(chat)
+    for playerId in itertools.chain(players_ids, [leader_id]):
+        player = session.query(Player).filter_by(id=playerId).first()
+        player.chat_id = chat.id
+        player.status = "CHATTING_AS_LEADER" if player.id == leader_id else "CHATTING_AS_PLAYER"
+    session.commit()
+    return chat.id
+
+def close(chat, started, winner_id):

флаг is_started есть в chat, зачем его передавать?



In server/chat_backend.py:
> +                is_closed=False, leader_id=leader_id, accepted=0)
+    session.add(chat)
+    for playerId in itertools.chain(players_ids, [leader_id]):
+        player = session.query(Player).filter_by(id=playerId).first()
+        player.chat_id = chat.id
+        player.status = "CHATTING_AS_LEADER" if player.id == leader_id else "CHATTING_AS_PLAYER"
+    session.commit()
+    return chat.id
+
+def close(chat, started, winner_id):
+    chat.is_closed = True
+    for player in chat.players:
+        player.chat_id = None
+        player.status = "IDLE"
+        if started and player.id != chat.leader_id:
+            player.rating = player.rating + (WINNER_DELTA if player.id == winner_id else LOSER_DELTA)

+=



In server/chat_backend.py:
> +    filename = whiteboard_location(base_filename)
+    with open(filename, "wb") as file:
+        file.write(whiteboard_body)
+    return base_filename
+
+
+def send_message(json):
+    if not json or 'authorId' not in json\
+            or 'text' not in json\
+            or 'chatId' not in json\
+            or not str(json["chatId"]).isdigit()\
+            or not str(json["authorId"]).isdigit():
+        return {}, 400
+    if json["text"] == "" and ('whiteboard' not in json 
+                               or json['whiteboard'] == ""):
+        return {}, 400

можно объединить эти два if'а, если ничего менять не планируется



In server/chat_backend.py:
> +    whiteboard_body = base64.b64decode(whiteboard_body)
+    base_filename = unique_filename()
+    filename = whiteboard_location(base_filename)
+    with open(filename, "wb") as file:
+        file.write(whiteboard_body)
+    return base_filename
+
+
+def send_message(json):
+    if not json or 'authorId' not in json\
+            or 'text' not in json\
+            or 'chatId' not in json\
+            or not str(json["chatId"]).isdigit()\
+            or not str(json["authorId"]).isdigit():
+        return {}, 400
+    if json["text"] == "" and ('whiteboard' not in json 

whiteboard что за штука? почему в сообщении должна передаваться?



In server/chat_backend.py:
> +    session.commit()
+    return message.toDict(), 200
+
+
+def get_messages(chat_id, num):
+    messages = session.query(Message).\
+                        filter_by(chat_id=chat_id).\
+                        order_by(Message.time).\
+                        offset(num).\
+                        all()
+    return list(map(Message.toDict, messages)), 200
+
+
+def get_leader(chat_id):
+    chat = session.query(Chat).filter_by(id=chat_id).first()
+    if not leader:

тоже сломано



In server/chat_backend.py:
> +
+
+def get_messages(chat_id, num):
+    messages = session.query(Message).\
+                        filter_by(chat_id=chat_id).\
+                        order_by(Message.time).\
+                        offset(num).\
+                        all()
+    return list(map(Message.toDict, messages)), 200
+
+
+def get_leader(chat_id):
+    chat = session.query(Chat).filter_by(id=chat_id).first()
+    if not leader:
+        return {"error": "Chat doesn't exist"}, 400
+    return {"leader": chat.leader_id}

200



In server/chat_backend.py:
> +                        all()
+    return list(map(Message.toDict, messages)), 200
+
+
+def get_leader(chat_id):
+    chat = session.query(Chat).filter_by(id=chat_id).first()
+    if not leader:
+        return {"error": "Chat doesn't exist"}, 400
+    return {"leader": chat.leader_id}
+
+
+def verify(chat):
+    if chat.is_closed or chat.is_started:
+        return
+    if (datetime.datetime.now() - chat.creation_time).total_seconds() > 25:
+        close(chat, False, -1)

из названия не очень понятно, зачем этот метод, что верифицирует и почему закрывает чат



In server/chat_backend.py:
> +        return {"error": "Player doesn't exist"}, 400
+    if not chat.is_closed and player not in chat.players:
+        return {"error": "You are not in required chat"}, 400
+    if chat.is_closed:
+        if not chat.is_started:
+            return {"result": "won't start"}, 200
+        elif chat.leader_id == player.id:
+            return {"result": "leader"}, 200
+        elif chat.winner_id == player.id:
+            return {"result": "winner", "rating": player.rating}, 200
+        else:
+            return {"result": "loser", "rating": player.rating}, 200
+    elif chat.is_started:
+        return {"result": "running"}, 200
+    else:
+        return {"result": "waiting"}, 200

не очень хорошо, что в метода смешался как статус чата, так и статус игрока в чате



In server/chat_backend.py:
> +        return {"result": "running"}, 200
+    else:
+        return {"result": "waiting"}, 200
+
+
+def accept(player_id):
+    print('accepted', player_id)
+    player = session.query(Player).filter_by(id=player_id).first()
+    if player is None:
+        return {"error": "Player doesn't exist"}, 400
+    if player.chat_id is None:
+        return {"error": "Player is not in chat"}, 400
+    chat = session.query(Chat).filter_by(id=player.chat_id).first()
+    verify(chat)
+   
+    chat.accepted += 1

аа, понял, это количество тех, кто "вступил" в чат. надо подумать над именем



In server/chat_backend.py:
> +    else:
+        return {"result": "waiting"}, 200
+
+
+def accept(player_id):
+    print('accepted', player_id)
+    player = session.query(Player).filter_by(id=player_id).first()
+    if player is None:
+        return {"error": "Player doesn't exist"}, 400
+    if player.chat_id is None:
+        return {"error": "Player is not in chat"}, 400
+    chat = session.query(Chat).filter_by(id=player.chat_id).first()
+    verify(chat)
+   
+    chat.accepted += 1
+    if chat.accepted == len(chat.players):

chat.players это что-то наподобие "приглашенных"?



In server/chat_backend.py:
> +    session.add(chat)
+    for playerId in itertools.chain(players_ids, [leader_id]):
+        player = session.query(Player).filter_by(id=playerId).first()
+        player.chat_id = chat.id
+        player.status = "CHATTING_AS_LEADER" if player.id == leader_id else "CHATTING_AS_PLAYER"
+    session.commit()
+    return chat.id
+
+def close(chat, started, winner_id):
+    chat.is_closed = True
+    for player in chat.players:
+        player.chat_id = None
+        player.status = "IDLE"
+        if started and player.id != chat.leader_id:
+            player.rating = player.rating + (WINNER_DELTA if player.id == winner_id else LOSER_DELTA)
+    session.commit()

чат могут закрыть несколько раз? рейтинг тогда каждый раз будет пересчитываться (



In server/chat_backend.py:
> +    return {}, 200
+
+def get_chat(chat_id):
+    chat = session.query(Chat).filter_by(id=chat_id).first()
+    return chat.toDict(), 200
+
+
+
+    
+
+def get_whiteboard(whiteboard_tag):
+    location = whiteboard_location(whiteboard_tag)
+    with open(location, "rb") as f:
+        whiteboard = f.read()
+        return base64.b64encode(whiteboard)
+    return ""

недостижимая инструкция



In server/battlemaker.py:
> @@ -0,0 +1,78 @@
+from db import Player, session
+from chat_backend import create_chat
+
+queue_filename = "queue.txt"
+player_amount = 1
+
+
+# Make sure file is present
+open(queue_filename, "a").close()
+
+
+def read_queues_from_file():
+    queues = []
+    with open(queue_filename, "r") as queue_file:
+        queues = [queue_file.readline()[:-1].split(",") for i in range(2)]

for _ in range(2)



In server/battlemaker.py:
> +# Make sure file is present
+open(queue_filename, "a").close()
+
+
+def read_queues_from_file():
+    queues = []
+    with open(queue_filename, "r") as queue_file:
+        queues = [queue_file.readline()[:-1].split(",") for i in range(2)]
+    queues = [list(map(int, filter(lambda s: str(s).isdigit(), queue)))
+              for queue in queues]
+    return queues[0], queues[1]
+
+
+def write_queues_in_file(players_queue, leaders_queue):
+    with open(queue_filename, "w") as queue_file:
+        queue_file.write(",".join(map(str, players_queue)) + '\n')

лучше строки не конкатенировать, а "\n" отдельно записать



In server/battlemaker.py:
> +    player = session.query(Player).filter_by(id=player_id).first()
+    if not player:
+        return {"error": "There's no such player."}, 400
+
+    if player.chat_id:
+        return {"error": "Player's already chatting."}, 400
+
+    players_queue, leaders_queue = read_queues_from_file()
+
+    if role == "player":
+        if player_id in players_queue:
+            return {"error": "Player's already in queue."}, 400
+        if player_id in leaders_queue:
+            leaders_queue.remove(player_id)
+        players_queue.append(player_id)
+        player.status = "IN_QUEUE_AS_PLAYER"

в питоне есть енамы, только не знаю, как их в бд класть



In server/battlemaker.py:
> +        for i in range(player_amount):
+            players.append(players_queue.pop(0))
+        create_chat(0, players, leaders_queue.pop(0))
+
+    write_queues_in_file(players_queue, leaders_queue)
+
+    return {}, 200
+
+
+def del_player_from_queue(player_id):
+    if not session.query(Player).filter_by(id=player_id).first():
+        return {"error": "There's no such player."}, 400
+
+    players_queue, leaders_queue = read_queues_from_file()
+
+    player = session.query(Player).filter_by(id=player_id).first()

один запрос производится два раза

—You are receiving this because you authored the thread.Reply to this email directly, view it on GitHub, or mute the thread.

Qwsafex committed:
Fixed images not displaying on load

sproshev commented on client/app/src/main/res/values-ru/strings.xml:
```diff
@@ -0,0 +1,49 @@
+    <string name="login_button">Войти</string>
+    <string name="signin_button">Войти</string>
+    <string name="signup_button">Зарегистрироваться</string>
+    <string name="signin_in">Вход...&#8230;</string>
```
шеститочие

sproshev commented on client/app/src/main/res/values-ru/strings.xml:
```diff
@@ -0,0 +1,49 @@
+    <string name="male">Мужской</string>
+    <string name="female">Женский</string>
+    <string name="password_dont_match">Пароли не совпадают</string>
+    <string name="signin_up">Регистрация...&#8230;</string>
```
шеститочие

sproshev commented on client/app/src/main/java/ru/spbau/shevchenko/chatbattle/Player.java:
```diff
@@ -0,0 +1,88 @@
+import org.json.JSONObject;
+
+public class Player {
+    // TODO: add fromJSON
```
todo устарело

sproshev commented on client/app/src/main/java/ru/spbau/shevchenko/chatbattle/Player.java:
```diff
@@ -0,0 +1,88 @@
+
+        public static Sex fromString(String sex) {
+            return Sex.valueOf(sex);
+        }
```
зачем этот метод тогда, если есть valueOf?

sproshev commented on client/app/src/main/java/ru/spbau/shevchenko/chatbattle/Player.java:
```diff
@@ -0,0 +1,88 @@
+        this.chatId = chatId;
+    }
+
+    private int chatId;
```
сущности надо упорядочить: поля, конструкторы, методы, внутр и влож классы

sproshev commented on client/app/src/main/java/ru/spbau/shevchenko/chatbattle/Message.java:
```diff
@@ -0,0 +1,59 @@
+import org.json.JSONObject;
+
+public class Message {
+    // TODO: add fromJSON
```
тоже устарело

sproshev commented on client/app/src/main/java/ru/spbau/shevchenko/chatbattle/Message.java:
```diff
@@ -0,0 +1,59 @@
+            return new Message(jsonMessage.getInt("id"), jsonMessage.getString("text"),
+                    jsonMessage.getInt("authorId"),
+                    jsonMessage.getInt("chatId"),
+                    (jsonMessage.isNull("whiteboardTag") ? "" : jsonMessage.getString("whiteboardTag")));
```
optString

sproshev commented on client/app/src/main/java/ru/spbau/shevchenko/chatbattle/Message.java:
```diff
@@ -0,0 +1,59 @@
+                    (jsonMessage.isNull("whiteboardTag") ? "" : jsonMessage.getString("whiteboardTag")));
+        } catch (JSONException e) {
+            Log.e("Message.fromJSON", "Failed to initialize message from JSON: " + e.getMessage());
+            throw e;
```
catch-блок не совпадает с Player

sproshev commented on client/app/src/main/java/ru/spbau/shevchenko/chatbattle/Message.java:
```diff
@@ -0,0 +1,59 @@
+        if (tag == null) {
+            tag = "";
+        }
+        this.tag = tag;
```
см тернарный оператор

sproshev commented on client/app/src/main/java/ru/spbau/shevchenko/chatbattle/frontend/WinnerPickActivity.java:
```diff
@@ -0,0 +1,63 @@
+                    Toast.makeText(WinnerPickActivity.this, R.string.choice_demand, Toast.LENGTH_LONG).show();
+                    break;
+                }
+                RequestMaker.chooseWinner(chosen);
```
про успешность или неуспешность отправки данных ничего не скажут, хотя бы про второе надо говорить

ZhNV2 replied:
Поддержка проблем с интернет-соединением будет в следующей версии

sproshev commented on client/app/src/main/java/ru/spbau/shevchenko/chatbattle/frontend/WhiteboardActivity.java:
```diff
@@ -0,0 +1,156 @@
+                        drawView.setBrushSize(largeBrush);
+                        brushDialog.dismiss();
+                    }
+                });
```
три раза один и тот же код --> метод?

sproshev commented on client/app/src/main/java/ru/spbau/shevchenko/chatbattle/frontend/WhiteboardActivity.java:
```diff
@@ -0,0 +1,156 @@
+                        brushDialog.dismiss();
+                    }
+                });
+                brushDialog.show();
```
и здесь

sproshev commented on client/app/src/main/java/ru/spbau/shevchenko/chatbattle/frontend/WhiteboardActivity.java:
```diff
@@ -0,0 +1,156 @@
+            }
+        }
+    }
+}
```
если пользователь просто нажмет назад, всё ок? проверяли?

ZhNV2 replied:
да, все окей, проверено

sproshev commented on client/app/src/main/java/ru/spbau/shevchenko/chatbattle/frontend/WhiteboardActivity.java:
```diff
@@ -0,0 +1,156 @@
+    //@RequiresApi(api = Build.VERSION_CODES.KITKAT)
+    @Override
+    public void onClick(View view) {
+        switch (view.getId()) {
```
такие длинные методы лучше разбивать на более мелкие, здесь можно разбить по типу нажатой кнопки

sproshev commented on client/app/src/main/java/ru/spbau/shevchenko/chatbattle/frontend/SearchActivity.java:
```diff
@@ -0,0 +1,139 @@
+        super.onCreate(savedInstanceState);
+        setContentView(R.layout.activity_search);
+
+        final Button search_as_player_button = (Button) findViewById(R.id.search_as_player_button);
```
именование

final на локальных переменных желательно писать либо везде, либо негде. для единого стиля

sproshev commented on client/app/src/main/java/ru/spbau/shevchenko/chatbattle/frontend/SearchActivity.java:
```diff
@@ -0,0 +1,139 @@
+        Log.d("Player status:", ProfileManager.getPlayerStatus().toString());
+        switch (ProfileManager.getPlayerStatus()) {
+            case IDLE: {
+                stop_searching();
```
почему если он уже IDLE, то надо дополнительно его удалить из очереди? он же наверное уже удален

ZhNV2 replied:
там же не только вызов delete_from_queue, я не очень хотел отдельно писать каждый раз и следить за тем, когда я уже послал сообщение, что надо его выкинуть

sproshev commented on client/app/src/main/java/ru/spbau/shevchenko/chatbattle/frontend/SearchActivity.java:
```diff
@@ -0,0 +1,139 @@
+                break;
+            }
+            case IN_QUEUE_AS_PLAYER: {
+                search_as(Player.Role.PLAYER);
```
если для игрока уже ищется батл, то для него еще батл начнут искать?

ZhNV2 replied:
Нет, если он вдруг не в очереди, то его заново добавит в очередь, иначе ничего не произойдет. Опять же, наверное, можно было отдельно каждый раз вызывать findBattle(), но не хотел так делать и следить за этим тоже

sproshev commented on client/app/src/main/java/ru/spbau/shevchenko/chatbattle/frontend/MessageAdapter.java:
```diff
@@ -0,0 +1,98 @@
+    public int add(Message message) {
+        messages.add(message);
+        notifyDataSetChanged();
+        return messages.size() - 1;
```
возвращаемое значение никем не используется

sproshev commented on client/app/src/main/java/ru/spbau/shevchenko/chatbattle/frontend/MessageAdapter.java:
```diff
@@ -0,0 +1,98 @@
+                }
+            });
+            if (whiteboardURI == null) {
+                holder.imageView.setImageResource(R.drawable.grey_square);
```
не знаю, сколько состояний может быть у whiteboard, типа "загружается", "отсутствует", "присутствует" или еще какие-то, но желательно отображать их все

это замечание с низким приоритетом, просто для удобства

sproshev commented on client/app/src/main/java/ru/spbau/shevchenko/chatbattle/frontend/MenuActivity.java:
```diff
@@ -0,0 +1,81 @@
+    public void onResume() {
+        super.onResume();
+        Class<?> previousClass = BasicActivity.getLastActivityClass();
+        if ((ProfileManager.getPlayer().getChatId() == -1) && previousClass != null && AbstractChat.class.isAssignableFrom(previousClass)) {
```
лишние скобки в первом равенстве

sproshev commented on client/app/src/main/java/ru/spbau/shevchenko/chatbattle/frontend/MenuActivity.java:
```diff
@@ -0,0 +1,81 @@
+        if ((ProfileManager.getPlayer().getChatId() == -1) && previousClass != null && AbstractChat.class.isAssignableFrom(previousClass)) {
+            Log.d("My app", previousClass.getName());
+            Intent intent = new Intent(this, SearchActivity.class);
+            startActivity(intent);
```
не уверен, что хорошая идея, может попробовать что-то со startActivityForResult

ZhNV2 replied:
не очень понял, про какую идею у вас возникли сомнения? 
Здесь реализовано следующее: если во время того, как баттл нашелся, вы в меню или searchActivity(), после баттла вас кидало в searchActivity()

>может попробовать что-то со startActivityForResult
Тоже не понял, что вы имеете в виду

sproshev commented on client/app/src/main/java/ru/spbau/shevchenko/chatbattle/frontend/LeaderActivity.java:
```diff
@@ -0,0 +1,56 @@
+        super.onCreate(savedInstanceState);
+        ProfileManager.setPlayerStatus(ProfileManager.PlayerStatus.CHATTING_AS_LEADER);
+    }
+
```
`@Override`

ZhNV2 replied:
не понял, он же перекрывает родительский onCreate(), в чем проблема?

sproshev commented on client/app/src/main/java/ru/spbau/shevchenko/chatbattle/frontend/DrawingView.java:
```diff
@@ -0,0 +1,136 @@
+    //canvas
+    private Canvas drawCanvas;
+    //canvas bitmap
+    private Bitmap canvasBitmap;
```
названия переменных почти совпадают с комментами, зачем тогда комменты?

sproshev commented on client/app/src/main/java/ru/spbau/shevchenko/chatbattle/frontend/BattleFoundDialogFragment.java:
```diff
@@ -0,0 +1,170 @@
+                })
+                .setNegativeButton(R.string.decline, new DialogInterface.OnClickListener() {
+                    public void onClick(DialogInterface dialog, int id) {
+                        RequestMaker.decline(ProfileManager.getPlayer().getId());
```
тоже желательно что-нибудь сказать, если не удалось. вообще это ко всему взаимодействию с сервером относится

sproshev commented on client/app/src/main/java/ru/spbau/shevchenko/chatbattle/frontend/BattleFoundDialogFragment.java:
```diff
@@ -0,0 +1,170 @@
+                        ProfileManager.setPlayerStatus(ProfileManager.PlayerStatus.IDLE);
+                        Activity activity = getActivity();
+                        if (activity instanceof SearchActivity) {
+                            ((SearchActivity) getActivity()).searchAgain(true, role);
```
activity

ZhNV2 replied:
не понял

sproshev commented on client/app/src/main/java/ru/spbau/shevchenko/chatbattle/frontend/BattleFoundDialogFragment.java:
```diff
@@ -0,0 +1,170 @@
+        builder.setMessage(R.string.battle_found)
+                .setPositiveButton(R.string.accept, new DialogInterface.OnClickListener() {
+                    public void onClick(DialogInterface dialog, int id) {
+                    }
```
почему слушатель этой кнопки реализован в другом месте?

ZhNV2 replied:
потому что иначе по нажатию на accept диалог закрывается

sproshev commented on client/app/src/main/java/ru/spbau/shevchenko/chatbattle/frontend/BattleFoundDialogFragment.java:
```diff
@@ -0,0 +1,170 @@
+                public boolean onKey(DialogInterface arg0, int keyCode, KeyEvent event) {
+                    return true;
+                }
+            });
```
это какой-то важный listener?  почему он только true возвращает?

ZhNV2 replied:
он запрещает нажимать кнопку назад

sproshev commented on client/app/src/main/java/ru/spbau/shevchenko/chatbattle/frontend/BattleFoundDialogFragment.java:
```diff
@@ -0,0 +1,170 @@
+
+    private long startTime;
+    private boolean hasAccepted;
+    final Handler getStatusHandler = new Handler();
```
private

sproshev commented on client/app/src/main/java/ru/spbau/shevchenko/chatbattle/frontend/BattleFoundDialogFragment.java:
```diff
@@ -0,0 +1,170 @@
+        }
+    };
+
+    final RequestCallback getStatusCallback = new RequestCallback() {
```
private

все поля описываются в начале класса

ZhNV2 replied:
Там же какая-то логика в них есть, все равно перенести?

sproshev commented on client/app/src/main/java/ru/spbau/shevchenko/chatbattle/frontend/BattleFoundDialogFragment.java:
```diff
@@ -0,0 +1,170 @@
+            try {
+                JSONObject playerObject = new JSONObject(response);
+                if (playerObject.has("error")) {
+                    return;
```
отобразить

sproshev commented on client/app/src/main/java/ru/spbau/shevchenko/chatbattle/frontend/BattleFoundDialogFragment.java:
```diff
@@ -0,0 +1,170 @@
+    final static private long HANDLER_DELAY = 100;
+    final private static int MAX_IDLENESS_TIME = 5000;
+
+    final Runnable getStatusRunnable = new Runnable() {
```
private

sproshev replied:
private

sproshev commented on client/app/src/main/java/ru/spbau/shevchenko/chatbattle/frontend/BattleFoundDialogFragment.java:
```diff
@@ -0,0 +1,170 @@
+                        if (currentActivity == null) return;
+                        if (hasAccepted || System.currentTimeMillis() - startTime < MAX_IDLENESS_TIME) {
+                            if (currentActivity instanceof SearchActivity) {
+                                ((SearchActivity) currentActivity).searchAgain(false, role);
```
за перезапуск поиска явно не активити отвечает, а сервис скорее

ZhNV2 replied:
Это не перезапуск поиска, а правильная настройка searchActivity, в который мы сейчас перейдем (там спиннер, правильную надпись и т.п.)

sproshev commented on client/app/src/main/java/ru/spbau/shevchenko/chatbattle/frontend/BattleFoundDialogFragment.java:
```diff
@@ -0,0 +1,170 @@
+                    }
+                }
+
+            } catch (JSONException e) {
```
слишком большой блок завернут в try-catch, а исключение бросается только в начале

sproshev commented on client/app/src/main/java/ru/spbau/shevchenko/chatbattle/frontend/BasicActivity.java:
```diff
@@ -0,0 +1,95 @@
+
+public class BasicActivity extends AppCompatActivity {
+
+    protected MyApplication myApplication;
```
private

sproshev commented on client/app/src/main/java/ru/spbau/shevchenko/chatbattle/frontend/BasicActivity.java:
```diff
@@ -0,0 +1,95 @@
+
+    public void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        myApplication = (MyApplication) this.getApplicationContext();
```
лишний this

sproshev commented on client/app/src/main/java/ru/spbau/shevchenko/chatbattle/frontend/BasicActivity.java:
```diff
@@ -0,0 +1,95 @@
+
+    @Override
+    public void onBackPressed() {
+        lastActivityClass = this.getClass();
```
this

sproshev commented on client/app/src/main/java/ru/spbau/shevchenko/chatbattle/frontend/BasicActivity.java:
```diff
@@ -0,0 +1,95 @@
+
+    @Override
+    public void onDestroy() {
+        lastActivityClass = this.getClass();
```
this

sproshev commented on client/app/src/main/java/ru/spbau/shevchenko/chatbattle/frontend/BasicActivity.java:
```diff
@@ -0,0 +1,95 @@
+
+    private void clearReferences() {
+        BasicActivity currActivity = myApplication.getCurrentActivity();
+        if (this.equals(currActivity))
```
this

sproshev commented on client/app/src/main/java/ru/spbau/shevchenko/chatbattle/frontend/BasicActivity.java:
```diff
@@ -0,0 +1,95 @@
+
+    final public static long BATTLE_FOUND_HANDLE_DELAY = 100;
+
+    public static class battleFoundRunnable implements Runnable {
```
классы именуются с большой буквы

sproshev commented on client/app/src/main/java/ru/spbau/shevchenko/chatbattle/frontend/AbstractChat.java:
```diff
@@ -0,0 +1,248 @@
+
+    private boolean initialized = false;
+
+    final protected ServiceConnection chatServiceConection = new ServiceConnection() {
```
private

sproshev commented on client/app/src/main/java/ru/spbau/shevchenko/chatbattle/frontend/AbstractChat.java:
```diff
@@ -0,0 +1,248 @@
+    private boolean initialized = false;
+
+    final protected ServiceConnection chatServiceConection = new ServiceConnection() {
+        @Override
```
и дальше

sproshev commented on client/app/src/main/java/ru/spbau/shevchenko/chatbattle/backend/SearcherService.java:
```diff
@@ -0,0 +1,150 @@
+            try {
+                Thread.sleep(200);
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
```
и что эта конструкция делает, почему из цикла не выходит?

sproshev commented on client/app/src/main/java/ru/spbau/shevchenko/chatbattle/backend/BattleSearcher.java:
```diff
@@ -0,0 +1,9 @@
+    public static void findBattle(final Player.Role role) {
+        RequestMaker.findBattle(role, ProfileManager.getPlayer().getId());
+    }
+}
```
класс своё отжил

sproshev reviewed (commented):
стало намного лучше

на что рекомендую обратить внимание:
обработка ошибок
код должен читаться максимально линейно, т.е. желательно, чтобы класс можно было прочитать за один проход без прыжков по нему

sproshev commented:
а, еще, рекомендую обращать внимание warning'и в идее и пайчарме

ZhNV2 committed:
cosmetic corrections from pr

ZhNV2 committed:
cosmetic corrections form pr (client)

ZhNV2 commented:
@sproshev Поправили все, что поняли. На остальное оставили комментарии

Qwsafex committed:
Changed RequestCallback.run parameter to RequestResult to pass request status. Added spinners to display message delivery process

Qwsafex committed:
Added timeout to sendRequest. Messages now turn red when there are internet connection problems.

ZhNV2 committed:
testing commit

Qwsafex committed:
Fixed bug in display of non-delivered messages

ZhNV2 committed:
testing features

ZhNV2 committed:
Implemented issue #19. Searcher Service was changed to Thread

ZhNV2 committed:
Implemented issue #27. SearcherThread was moved to LogicActivity. Normal spinner item's size in menu

ZhNV2 committed:
cosmetic changes in design

Qwsafex committed:
Working on internet connection troubles handling...

ZhNV2 committed:
Cosmetic changes in design. New class for using translated constants in code

ZhNV2 committed:
light version of changing password feature is implemented

Qwsafex committed:
Basic internet connection troubles handling (issue #9)

Qwsafex committed:
Issue #9 merged

ZhNV2 committed:
Normal change password feature

Qwsafex committed:
Smth changed

ZhNV2 committed:
normal version of password changing

ZhNV2 committed:
upper

Qwsafex committed:
Hot fixes =)

ZhNV2 committed:
new design of login activity

ZhNV2 committed:
Merge branch 'new-features' of https://github.com/BatherGaming/chat-battle into new-features

ZhNV2 committed:
new sign up design

ZhNV2 committed:
navigation drawer in menu

ZhNV2 committed:
profile activity was deleted

ZhNV2 committed:
design fixes

ZhNV2 committed:
design fixes

ZhNV2 committed:
search activity design

ZhNV2 committed:
desgin fixes

Qwsafex committed:
Added simple leaderboard

ZhNV2 committed:
chat design

Qwsafex committed:
Working on leaderboard

Qwsafex committed:
Removed redundant fields from SignupActivity

Qwsafex committed:
Merge branch 'feat/leaderboard' into new-features

ZhNV2 committed:
password recovery

Qwsafex committed:
Added battle timer

Qwsafex committed:
Merged

ZhNV2 committed:
some mistakes were fixed

ZhNV2 committed:
basic ui elements

Qwsafex committed:
Working on battle summary

Qwsafex committed:
Merged

Qwsafex committed:
Fixes

ZhNV2 committed:
buildgradle

ZhNV2 committed:
Merge branch 'new-features' of https://github.com/BatherGaming/chat-battle into new-features

ZhNV2 committed:
design in login,singup, menu

Qwsafex committed:
Semifinished summary

Qwsafex committed:
Added some translations

ZhNV2 committed:
search activity design

Qwsafex committed:
Leaderboard done

Qwsafex committed:
Deleted redundant xml file

Qwsafex committed:
Chat-battle title design fix

ZhNV2 committed:
chat design

Qwsafex committed:
Drawable file structure reorganized and summary finalized

Qwsafex committed:
Merged

Qwsafex committed:
Small fixes

Qwsafex committed:
First version of battle list

ZhNV2 committed:
chat design

Qwsafex committed:
First version of spectate feature

ZhNV2 committed:
chat design

Qwsafex committed:
Small changes

Qwsafex committed:
Merged

ZhNV2 committed:
leaderboard

Qwsafex committed:
Messages consisting only of whitespace characters are prohibited now

Qwsafex committed:
PEP8 style fixes

Qwsafex committed:
Server methods renamed and some minor refactoring done

Qwsafex committed:
Test

Qwsafex committed:
Undo test

Qwsafex committed:
Minor response code changes

ZhNV2 committed:
layout refactoring

ZhNV2 committed:
redundant drawables and layouts were deleted. All resourses are refactored

Qwsafex committed:
Little refactoring

Qwsafex committed:
Refactored

ZhNV2 committed:
refactoring

ZhNV2 committed:
ref

ZhNV2 committed:
refactoring

ZhNV2 committed:
refactoring

Qwsafex committed:
Refactored

Qwsafex committed:
Merged

Qwsafex committed:
Refactored

ZhNV2 committed:
ref + navigationView in all activities

ZhNV2 committed:
ref + navDrawer in all activies

Qwsafex committed:
Merged

ZhNV2 committed:
battle list

Qwsafex committed:
Some bugs fixed and whiteboard design half-done

Qwsafex committed:
Whiteboard design finished and ref

Qwsafex committed:
Whiteboard now persists on comeback to WhiteboardActivity

ZhNV2 committed:
fixes

ZhNV2 committed:
ref

Qwsafex committed:
Fixed player able to send messages while spectating if he was in another chat

Qwsafex committed:
fixes

Qwsafex committed:
Drawer now closes when button with same activity clicked

Qwsafex committed:
Removed debug output

Qwsafex committed:
Incorrect battle list order fixed

Qwsafex committed:
Multiple MenuActivities on login fixed

Qwsafex committed:
Bug fixed

Qwsafex committed:
Translations fixed
